File: TeamCode/src/main/java/org/riverdell/robotics/xdk/opmodes/game/DevTeleOp.java
package org.riverdell.robotics.xdk.opmodes.game;

import com.qualcomm.robotcore.eventloop.opmode.TeleOp;
import io.liftgate.ftc.scripting.opmode.DevLinearOpMode;
import io.liftgate.robotics.mono.Mono;
import io.liftgate.robotics.mono.gamepad.ButtonType;
import io.liftgate.robotics.mono.gamepad.GamepadCommands;
import kotlin.Unit;
import org.jetbrains.annotations.NotNull;

import java.util.Collections;
import java.util.List;

/**
 * @author GrowlyX
 * @since 9/5/2023
 */
@TeleOp(name = "Dev", group = "game")
public class DevTeleOp extends DevLinearOpMode {

    @Override
    public void runOpMode() {
        buildCommands();
        super.runOpMode();
    }

    private void buildCommands() {
        final GamepadCommands gp1Commands = Mono
            .INSTANCE.commands(gamepad1);

        gp1Commands.where(ButtonType.ButtonX)
            .triggers(() -> {
                System.out.println("hey :)");
                return Unit.INSTANCE;
            })
            .whenPressedOnce();

        gp1Commands.startListening();
    }

    @Override
    public @NotNull String getScriptName() {
        return "Dev.ts";
    }

    @Override
    public @NotNull List<ImpliedVariable> getImpliedVariables() {
        return Collections.emptyList();
    }
}

File: TeamCode/src/main/kotlin/org/riverdell/robotics/opmode/DevTeleOp.kt
package org.riverdell.robotics.opmode

import com.qualcomm.robotcore.eventloop.opmode.TeleOp
import io.liftgate.ftc.scripting.KotlinScript.ImpliedVariable
import io.liftgate.ftc.scripting.opmode.DevLinearOpMode
import org.riverdell.robotics.gamepad.ButtonType
import org.riverdell.robotics.gamepad.GamepadCommands

@TeleOp(
    name = "kts - Dev Test",
    group = "scripted"
)
class DevTeleOp : DevLinearOpMode()
{
    private val gp1Commands by lazy { GamepadCommands(gamepad1) }
    private val gp2Commands by lazy { GamepadCommands(gamepad2) }

    fun configureCommands()
    {
        // TODO: add hooks to ProdOpMode for prestart conf stuff, etc
        listOf(gp1Commands, gp2Commands)
            .forEach(GamepadCommands::startListening)

        fun toggleLiftMotorSlowMode() = Unit

        gp1Commands
            .where(ButtonType.ButtonA)
            .triggers(::toggleLiftMotorSlowMode) // toggle each time the user presses A once
            .whenPressedOnce()

        fun toggleHoldingOfLiftMotor() = Unit

        gp1Commands
            .where(ButtonType.ButtonX)
            .triggers(::toggleHoldingOfLiftMotor) // toggle lift motor hold when user starts holding X
            // maintain lift motor hold
            .andIsHeldUntilReleasedWhere(::toggleHoldingOfLiftMotor) // when user stops holding X, disable lift hold

        fun bringUpLiftMotor() = Unit

        gp1Commands
            .where(ButtonType.ButtonY)
            .and(ButtonType.ButtonB)
            .triggers(::bringUpLiftMotor, 250L) // continue to increment lift motor position while the user is holding both B and Y every 250ms
            .repeatedlyWhilePressed()

        fun resetLiftMotorPosition() = Unit
        fun inEndGame(): Boolean = false

        // Similar example to the one right above this, however, this time we are able to
        // reset the lift motor back to position 0 when the user stops pressing
        gp1Commands
            .where(ButtonType.ButtonY)
            .and(ButtonType.ButtonB) // can either use B + Y
            .or {
                and(ButtonType.DPadUp)
                    .and(ButtonType.DPadLeft)
            } // or they can use DPUp + DPLeft
            .onlyWhen(::inEndGame) // only allow this command when in end game
            .triggers(::bringUpLiftMotor, 500L) // continue to call this ever 50ms while the user is holding both B and Y every 500ms
            .repeatedlyWhilePressedUntilReleasedWhere(::resetLiftMotorPosition)
    }

    override fun runOpMode()
    {
        configureCommands()
        super.runOpMode()
    }

    override fun getImpliedVariables() = mutableListOf(
        ImpliedVariable.of("gp1Commands", gp1Commands),
        ImpliedVariable.of("gp2Commands", gp2Commands)
    )
    override fun getScriptName() = "Dev.kts"
}
